### PR TITLE
refactor(pubsub): make room for Subscription retry

### PR DESCRIPTION
**Pull request by coryan**
A new class implementing the `SubscriptionBatchSource` interface. This
class will later implement the retry loops, but first refactor
everything so the PRs are easier to grok.

Part of the work for #4699

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5017)

### CODE REVIEW

File: google/cloud/pubsub/internal/default_subscription_batch_source.cc [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+  google::pubsub::v1::ModifyAckDeadlineRequest request;
+  request.set_subscription(subscription_full_name_);
+  for (auto& id : ack_ids) {
+    request.add_ack_ids(std::move(id));
+  }
+  request.set_ack_deadline_seconds(0);
+  return stub_->AsyncModifyAckDeadline(
+      cq_, absl::make_unique<grpc::ClientContext>(), request);
+}
+future<Status> DefaultSubscriptionBatchSource::ExtendLeases(
```
Comment (devbww) [inline google/cloud/pubsub/internal/default_subscription_batch_source.cc:54]: Insert empty line.

File: google/cloud/pubsub/internal/default_subscription_batch_source.cc [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    std::vector<std::string> ack_ids, std::chrono::seconds extension) {
+  google::pubsub::v1::ModifyAckDeadlineRequest request;
+  request.set_subscription(subscription_full_name_);
+  request.set_ack_deadline_seconds(
+      static_cast<std::int32_t>(extension.count()));
```
Comment (devbww) [inline google/cloud/pubsub/internal/default_subscription_batch_source.cc:59]: Opt: Perhaps we should be saturating the `int32_t` in overflow cases?

**Review by devbww — APPROVED** [review]

File: google/cloud/pubsub/internal/subscription_lease_management.cc [hunk 3]
```diff
@@ -29,57 +29,46 @@ void SubscriptionLeaseManagement::Shutdown() {
...
-  for (auto& id : ack_ids) {
-    request.add_ack_ids(std::move(id));
-  }
-  request.set_ack_deadline_seconds(0);
-  return stub_->AsyncModifyAckDeadline(
-      cq_, absl::make_unique<grpc::ClientContext>(), request);
+  return child_->BulkNack(std::move(ack_ids), total_size);
+}
+
+// Users of this class should have no need to call ExtendLeases(), they create
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/pubsub/internal/subscription_lease_management.cc]: Done.

File: google/cloud/pubsub/internal/default_subscription_batch_source.cc (hunk 1 above)
Comment (coryan) [reply google/cloud/pubsub/internal/default_subscription_batch_source.cc:55]: Fixed, thanks.

**Review by devbww — APPROVED** [review]

**Pull request merged**